### PR TITLE
Take our modified xformers microGPT script and add Profiling

### PR DESCRIPTION
**Pull request by SeanNaren**

### CODE REVIEW

**Comment by SeanNaren** [comment]
I've added a reference pytorch only script (still using the lightning seed everything function) to just ensure that pytorch lightning doesn't add a ridiculous overhead or something. From my initial experiments on a 3090, seems cool!

Waiting for A100 resources next week to be able to really test this out. In the meantime, going to look a bit further into some suggestions by Ben :)

File: train.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+                block_size=block_size,
+                vocab_size=train_dataset.vocab_size,
+                activation_checkpointing=False
+            ),
+            CUDAMemoryCallback(),
+        ],
+        limit_train_batches=50,
+        max_epochs=epochs,
+        precision=precision,
+        # gradient_clip_val=1,  # Use to catch divergent gradients, if experimenting
```
Comment (SeanNaren) [inline train.py:302]:
```suggestion
        gradient_clip_val=1,  # Use to catch divergent gradients, if experimenting
```
